# Learning Objective

- Solve basic questions using
  - Numpy
  - Pandas
- Work on a case study using numpy and pandas

## Motivation 
We have seen python basic data structures in our last class. They are great but lack specialized features for data analysis. Like, adding roows, columns, operating on 2d matrices aren't readily available. So, we will use *numpy* and *pandas* for such functions.

## Why numpy and pandas are so popular?
- Provide high-performance, easy-to-use data structures and data analysis tools for the Python
- Have many advantages over Python inbuilt data structures
- Have many inbuilt efficient functions and hence consume less memory and provide fast computation
- Have become must for Data Science due to their many advantages for data analysis, data cleaning, data preparation, etc.
 


## Numpy

In [ ]:
import numpy as np #import numpy

### Ques 1. Create two  arrays of 5*5 (i.e. 5 rows and 5 columns) using first 25 elements (i.e. 0 to 24). 
### Calculate the inner product of the two arrays using numpy.


In [ ]:
arr_1 = np.arange(25).reshape(5,5)
arr_2 = np.arange(25).reshape(5,5)

print(arr_1)

print(arr_2)

[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]]
[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]]


In [ ]:
import time
start_time = time.time()
print(np.inner(arr_1,arr_2))
print("--- %s seconds ---" % (time.time() - start_time))

[[  30   80  130  180  230]
 [  80  255  430  605  780]
 [ 130  430  730 1030 1330]
 [ 180  605 1030 1455 1880]
 [ 230  780 1330 1880 2430]]
--- 0.0025687217712402344 seconds ---


- A very complex computation executed withing seconds

### Ques 2: There are two array `x` and `y`. Write a program to output the matching items in both the array.
    
- x = np.array([1,2,3,4,5])
- y = np.array([1,6,7,8, 5, 9])
- Output = [1,5]

In [ ]:
import numpy as np
x = np.array([1,2,3,4,5])
y = np.array([1,6,7,8,5,9])

np.intersect1d(x, y)

array([1, 5])

- Can we write the same code using loops?

### Ques 3. Write a program to sort a given array of shape 2 along the first axis, last axis and on flattened array.
a = np.array([[10,40],[30,20]])

In [ ]:
a = np.array([[10,40],[30,20]])
print("Original array:")
print(a)
print("Sort the array along the first axis:")
print(np.sort(a, axis=0))
print("Sort the array along the last axis:")
print(np.sort(a))
print("Sort the flattened array:")
print(np.sort(a, axis=None))

Original array:
[[10 40]
 [30 20]]
Sort the array along the first axis:
[[10 20]
 [30 40]]
Sort the array along the last axis:
[[10 40]
 [20 30]]
Sort the flattened array:
[10 20 30 40]


## Pandas


### Ques 1. Convert the below two dictionaries to two dataframes. And concatenate these twoo DFs into one dataframe.

 - dict_1={'emp_id':['emp1','emp2','emp3','emp4'],'name':['Adam','Ellie','Rose','Slash'],'Ratings':[3,5,2,1]}
 - dict_2={'emp_id':['emp5','emp6','emp7','emp8'],'name':['Ram','Hari','Gopal','Shreyas'],'Ratings':[4,1,2,3]}

In [ ]:
dict_1={'emp_id':['emp1','emp2','emp3','emp4'],'name':['Adam','Ellie','Rose','Slash'],'Ratings':[3,5,2,1]}
dict_2={'emp_id':['emp5','emp6','emp7','emp8'],'name':['Ram','Hari','Gopal','Shreyas'],'Ratings':[4,1,2,3]}

df1=pd.DataFrame(dict_1)
df2=pd.DataFrame(dict_2)

final_df=pd.concat([df1,df2])

final_df

emp_id     name  Ratings
0   emp1     Adam        3
1   emp2    Ellie        5
2   emp3     Rose        2
3   emp4    Slash        1
0   emp5      Ram        4
1   emp6     Hari        1
2   emp7    Gopal        2
3   emp8  Shreyas        3

### Ques 2. Given the below three dictionaries. 
### 2.1 Convert these dictionaries into dataframes
### 2.2 Concatenate the first two DFs into one DF
### 2.3 Merge the above combined DF with the coupon DF
emp_data1 = pd.DataFrame({
        'emp_id': ['emp1', 'emp2', 'emp3', 'emp4', 'emp5'],
         'name': ['Danniella', 'Ryder', 'Bryce', 'Ed', 'Kwame'], 
        'salary': [200, 210, 190, 222, 199]})

emp_data2 = pd.DataFrame({
        'emp_id': ['emp4', 'emp5', 'emp6', 'emp7', 'emp8'],
        'name': ['Scarlette', 'Carla', 'Dante', 'Kaiser', 'Madeeha'], 
        'salary': [201, 200, 198, 219, 201]})

coupon_data = pd.DataFrame({
        'emp_id': ['emp1', 'emp2', 'emp3', 'emp4', 'emp5', 'emp7', 'emp8', 'emp9', 'emp10', 'emp11', 'emp12', 'emp13'],
        'cupon_id': [23, 45, 12, 67, 21, 55, 33, 14, 56, 83, 88, 12]})

In [ ]:
import pandas as pd
emp_data1 = pd.DataFrame({
        'emp_id': ['emp1', 'emp2', 'emp3', 'emp4', 'emp5'],
         'name': ['Danniella', 'Ryder', 'Bryce', 'Ed', 'Kwame'], 
        'salary': [200, 210, 190, 222, 199]})

emp_data2 = pd.DataFrame({
        'emp_id': ['emp4', 'emp5', 'emp6', 'emp7', 'emp8'],
        'name': ['Scarlette', 'Carla', 'Dante', 'Kaise', 'Madeeha'], 
        'salary': [201, 200, 198, 219, 201]})

coupon_data = pd.DataFrame({
        'emp_id': ['emp1', 'emp2', 'emp3', 'emp4', 'emp5', 'emp7', 'emp8', 'emp9', 'emp10', 'emp11', 'emp12', 'emp13'],
        'cupon_id': [23, 45, 12, 67, 21, 55, 33, 14, 56, 83, 88, 12]})

In [ ]:
print("Original DataFrames:")
print(emp_data1)
print("*"*50)
print(emp_data2)
print("*"*50)
print(coupon_data)

Original DataFrames:
  emp_id       name  salary
0   emp1  Danniella     200
1   emp2      Ryder     210
2   emp3      Bryce     190
3   emp4         Ed     222
4   emp5      Kwame     199
**************************************************
  emp_id       name  salary
0   emp4  Scarlette     201
1   emp5      Carla     200
2   emp6      Dante     198
3   emp7      Kaise     219
4   emp8    Madeeha     201
**************************************************
   emp_id  cupon_id
0    emp1        23
1    emp2        45
2    emp3        12
3    emp4        67
4    emp5        21
5    emp7        55
6    emp8        33
7    emp9        14
8   emp10        56
9   emp11        83
10  emp12        88
11  emp13        12


In [ ]:
print("\nJoin first two said dataframes along rows:")
result_data = pd.concat([emp_data1, emp_data2])
print(result_data)


Join first two said dataframes along rows:
  emp_id       name  salary
0   emp1  Danniella     200
1   emp2      Ryder     210
2   emp3      Bryce     190
3   emp4         Ed     222
4   emp5      Kwame     199
0   emp4  Scarlette     201
1   emp5      Carla     200
2   emp6      Dante     198
3   emp7      Kaise     219
4   emp8    Madeeha     201


In [ ]:
print("\nNow join the said result_data and coupon_data along emp_id:")
final_merged_data = pd.merge(result_data, coupon_data, on='emp_id')
print(final_merged_data)


Now join the said result_data and coupon_data along emp_id:
  emp_id       name  salary  cupon_id
0   emp1  Danniella     200        23
1   emp2      Ryder     210        45
2   emp3      Bryce     190        12
3   emp4         Ed     222        67
4   emp4  Scarlette     201        67
5   emp5      Kwame     199        21
6   emp5      Carla     200        21
7   emp7      Kaise     219        55
8   emp8    Madeeha     201        33


### Ques 3. Check 5 random samples of the above DF. Check the shape and info of the above DataFrame.

In [ ]:
final_merged_data.sample(5)

emp_id       name  salary  cupon_id
7   emp7      Kaise     219        55
1   emp2      Ryder     210        45
6   emp5      Carla     200        21
4   emp4  Scarlette     201        67
2   emp3      Bryce     190        12

In [ ]:
final_merged_data.shape

(9, 4)

In [ ]:
final_merged_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9 entries, 0 to 8
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   emp_id    9 non-null      object
 1   name      9 non-null      object
 2   salary    9 non-null      int64 
 3   cupon_id  9 non-null      int64 
dtypes: int64(2), object(2)
memory usage: 360.0+ bytes


## Case Study

For the below set of questions, we will be using the famous titanic disaster dataset.
The dataset contains,
- survived -   whether the passenger has survived or not 0 = No, 1 = Yes
- pclass   -  Class of the passenger
- sex      -   Gender of the passenger
- age      -   Age of the passenger
- sibsp    -   # of siblings / spouses aboard the Titanic
- parch    -   # of parents / children aboard the Titanic
- fare     -   Passenger fare
- embarked -   Port of Embarkation like: C = Cherbourg, Q = Queenstown, S =        Southampton
- deck     -   Deck data

### Ques 1. Load your dataset

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import os
os.chdir('/content/drive/MyDrive/')

In [1]:
import pandas as pd
df=pd.read_csv('titanic.csv')

### Ques 2. Check the first 5 lines and last 5 lines of the above dataframe

Hint: use .head() and .tail()



In [2]:
df.head(5)

survived  pclass     sex   age  sibsp  parch     fare embarked deck
0         0       3    male  22.0      1      0   7.2500        S  NaN
1         1       1  female  38.0      1      0  71.2833        C    C
2         1       3  female  26.0      0      0   7.9250        S  NaN
3         1       1  female  35.0      1      0  53.1000        S    C
4         0       3    male  35.0      0      0   8.0500        S  NaN

In [3]:
df.tail(5)

survived  pclass     sex   age  sibsp  parch   fare embarked deck
886         0       2    male  27.0      0      0  13.00        S  NaN
887         1       1  female  19.0      0      0  30.00        S    B
888         0       3  female   NaN      1      2  23.45        S  NaN
889         1       1    male  26.0      0      0  30.00        C    C
890         0       3    male  32.0      0      0   7.75        Q  NaN

### Ques 3. Check the basic statistics of the columns. Check missing values

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  891 non-null    int64  
 1   pclass    891 non-null    int64  
 2   sex       891 non-null    object 
 3   age       714 non-null    float64
 4   sibsp     891 non-null    int64  
 5   parch     891 non-null    int64  
 6   fare      891 non-null    float64
 7   embarked  889 non-null    object 
 8   deck      203 non-null    object 
dtypes: float64(2), int64(4), object(3)
memory usage: 62.8+ KB


In [5]:
df.isnull().sum()

survived      0
pclass        0
sex           0
age         177
sibsp         0
parch         0
fare          0
embarked      2
deck        688
dtype: int64

We will discuss how to deal with missing data in upcoming modules.

### Ques 4. Use describe method to get the basic statistics of both continuous and categorical features.

In [6]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [8]:
df.describe(include=['O'])

sex embarked deck
count    891      889  203
unique     2        3    7
top     male        S    C
freq     577      644   59

- Sex variable has two possible values with 65% male (top=male, freq=577/count=891).
- Embarked takes three possible values. S port used by most passengers (top=S)
- Deck takes seven unique values. C deck is used by most passengers (top=C)

### Ques 5. Check correlation amongst features using corr() function of pandas dataframe

In [9]:
df.corr()

survived    pclass       age     sibsp     parch      fare
survived  1.000000 -0.338481 -0.077221 -0.035322  0.081629  0.257307
pclass   -0.338481  1.000000 -0.369226  0.083081  0.018443 -0.549500
age      -0.077221 -0.369226  1.000000 -0.308247 -0.189119  0.096067
sibsp    -0.035322  0.083081 -0.308247  1.000000  0.414838  0.159651
parch     0.081629  0.018443 -0.189119  0.414838  1.000000  0.216225
fare      0.257307 -0.549500  0.096067  0.159651  0.216225  1.000000

### Ques 6. Report how many female passengers died? (Hint:- use inbuilt functions)

In [10]:
pd.crosstab(df['sex'],df['survived'])

survived    0    1
sex               
female     81  233
male      468  109

********************************************************************************

### Ques 7. Make a sub dataframe which consists of only sex and age column. Print the mean age of male and female using groupby function.

In [11]:
sub_df=df.iloc[:,[2,3]]

 Use groupby to display mean of both male and female.

In [12]:
sub_df.groupby("sex").mean()

age
sex              
female  27.915709
male    30.726645

********************************************************************************

### Ques 8. Print the maximum value of age and fare for male and female separately using groupby.
Outcome should have max age of female, max age of male, max fare of feamle, and max fare of male.

In [13]:
df.groupby("sex")[["fare","age"]].max()

fare   age
sex                   
female  512.3292  63.0
male    512.3292  80.0

********************************************************************************

### Ques 9. What is the average rate of survived passengers in different pclasses?

In [14]:
df[['pclass', 'survived']].groupby(['pclass'], as_index=False).mean().sort_values(by='survived', ascending=False)

pclass  survived
0       1  0.629630
1       2  0.472826
2       3  0.242363

********************************************************************************

### Ques 10. What is the average survival rate of male and female passengers ?

In [15]:
df[["sex", "survived"]].groupby(['sex'], as_index=False).mean().sort_values(by='survived', ascending=False)

sex  survived
0  female  0.742038
1    male  0.188908

We can see that survival rate of female passengers is higher(~74%).

********************************************************************************

### Ques 11. Print average survived passengers based on their total number of siblings/spouse who aboard the titanic ship (Hint : use groupby over sibsp feature)

In [16]:
df[["sibsp", "survived"]].groupby(['sibsp'], as_index=False).mean().sort_values(by='survived', ascending=False)

sibsp  survived
1      1  0.535885
2      2  0.464286
0      0  0.345395
3      3  0.250000
4      4  0.166667
5      5  0.000000
6      8  0.000000

We can conclude that passengers with only single sibling/spouse aboarded, had highest percentage of survival.(i.e. 53.5%)

********************************************************************************

### Ques 12. Print average survived passengers based on their number of parents/children who aboard the titanic ship.

In [17]:
df[["parch", "survived"]].groupby(['parch'], as_index=False).mean().sort_values(by='survived', ascending=False)

parch  survived
3      3  0.600000
1      1  0.550847
2      2  0.500000
0      0  0.343658
5      5  0.200000
4      4  0.000000
6      6  0.000000

We can conclude that passengers with 3 parents/children aboarded, had highest percentage of survival.(i.e. 60%)

********************************************************************************

### Ques 13. Calculate adult passengers( age > 18) aboarded. (Use lambda function)

In [18]:
df[df["age"].apply(lambda x: x > 18)]

survived  pclass     sex   age  sibsp  parch     fare embarked deck
0           0       3    male  22.0      1      0   7.2500        S  NaN
1           1       1  female  38.0      1      0  71.2833        C    C
2           1       3  female  26.0      0      0   7.9250        S  NaN
3           1       1  female  35.0      1      0  53.1000        S    C
4           0       3    male  35.0      0      0   8.0500        S  NaN
..        ...     ...     ...   ...    ...    ...      ...      ...  ...
885         0       3  female  39.0      0      5  29.1250        Q  NaN
886         0       2    male  27.0      0      0  13.0000        S  NaN
887         1       1  female  19.0      0      0  30.0000        S    B
889         1       1    male  26.0      0      0  30.0000        C    C
890         0       3    male  32.0      0      0   7.7500        Q  NaN

[575 rows x 9 columns]

Total 752 passengers are adults.

********************************************************************************

### Ques 14. Calculate the total number of passengers who do not have any siblings/spouse on board.

In [19]:
df[df["sibsp"].apply(lambda x: x == 0)]

survived  pclass     sex   age  sibsp  parch     fare embarked deck
2           1       3  female  26.0      0      0   7.9250        S  NaN
4           0       3    male  35.0      0      0   8.0500        S  NaN
5           0       3    male   NaN      0      0   8.4583        Q  NaN
6           0       1    male  54.0      0      0  51.8625        S    E
8           1       3  female  27.0      0      2  11.1333        S  NaN
..        ...     ...     ...   ...    ...    ...      ...      ...  ...
885         0       3  female  39.0      0      5  29.1250        Q  NaN
886         0       2    male  27.0      0      0  13.0000        S  NaN
887         1       1  female  19.0      0      0  30.0000        S    B
889         1       1    male  26.0      0      0  30.0000        C    C
890         0       3    male  32.0      0      0   7.7500        Q  NaN

[608 rows x 9 columns]

********************************************************************************

### Ques 15. Display the data for infants who survived ( infants where age <=4)

In [20]:
df.loc[df["survived"] == 1][df["age"].apply(lambda x: x <= 4)]


/var/folders/nt/3mk_43xn04n4h7tyjmwhy93w0000gp/T/ipykernel_1575/1221741481.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df.loc[df["survived"] == 1][df["age"].apply(lambda x: x <= 4)]


survived  pclass     sex   age  sibsp  parch      fare embarked deck
10          1       3  female  4.00      1      1   16.7000        S    G
43          1       2  female  3.00      1      2   41.5792        C  NaN
78          1       2    male  0.83      0      2   29.0000        S  NaN
172         1       3  female  1.00      1      1   11.1333        S  NaN
183         1       2    male  1.00      2      1   39.0000        S    F
184         1       3  female  4.00      0      2   22.0250        S  NaN
193         1       2    male  3.00      1      1   26.0000        S    F
261         1       3    male  3.00      4      2   31.3875        S  NaN
305         1       1    male  0.92      1      2  151.5500        S    C
340         1       2    male  2.00      1      1   26.0000        S    F
348         1       3    male  3.00      1      1   15.9000        S  NaN
381         1       3  female  1.00      0      2   15.7417        C  NaN
407         1       2    male  3.00      1      1   18.7500        S  NaN
445         1       1    male  4.00      0      2   81.8583        S    A
469         1       3  female  0.75      2      1   19.2583        C  NaN
479         1       3  female  2.00      0      1   12.2875        S  NaN
530         1       2  female  2.00      1      1   26.0000        S  NaN
618         1       2  female  4.00      2      1   39.0000        S    F
644         1       3  female  0.75      2      1   19.2583        C  NaN
691         1       3  female  4.00      0      1   13.4167        C  NaN
750         1       2  female  4.00      1      1   23.0000        S  NaN
755         1       2    male  0.67      1      1   14.5000        S  NaN
788         1       3    male  1.00      1      2   20.5750        S  NaN
803         1       3    male  0.42      0      1    8.5167        C  NaN
827         1       2    male  1.00      0      2   37.0042        C  NaN
831         1       2    male  0.83      1      1   18.7500        S  NaN
869         1       3    male  4.00      1      1   11.1333        S  NaN

OR we can also check it using logical and (&) condition over dataframe

In [21]:
surviving_children = df[(df["age"] <= 4) & (df["survived"] == 1)]
surviving_children

survived  pclass     sex   age  sibsp  parch      fare embarked deck
10          1       3  female  4.00      1      1   16.7000        S    G
43          1       2  female  3.00      1      2   41.5792        C  NaN
78          1       2    male  0.83      0      2   29.0000        S  NaN
172         1       3  female  1.00      1      1   11.1333        S  NaN
183         1       2    male  1.00      2      1   39.0000        S    F
184         1       3  female  4.00      0      2   22.0250        S  NaN
193         1       2    male  3.00      1      1   26.0000        S    F
261         1       3    male  3.00      4      2   31.3875        S  NaN
305         1       1    male  0.92      1      2  151.5500        S    C
340         1       2    male  2.00      1      1   26.0000        S    F
348         1       3    male  3.00      1      1   15.9000        S  NaN
381         1       3  female  1.00      0      2   15.7417        C  NaN
407         1       2    male  3.00      1      1   18.7500        S  NaN
445         1       1    male  4.00      0      2   81.8583        S    A
469         1       3  female  0.75      2      1   19.2583        C  NaN
479         1       3  female  2.00      0      1   12.2875        S  NaN
530         1       2  female  2.00      1      1   26.0000        S  NaN
618         1       2  female  4.00      2      1   39.0000        S    F
644         1       3  female  0.75      2      1   19.2583        C  NaN
691         1       3  female  4.00      0      1   13.4167        C  NaN
750         1       2  female  4.00      1      1   23.0000        S  NaN
755         1       2    male  0.67      1      1   14.5000        S  NaN
788         1       3    male  1.00      1      2   20.5750        S  NaN
803         1       3    male  0.42      0      1    8.5167        C  NaN
827         1       2    male  1.00      0      2   37.0042        C  NaN
831         1       2    male  0.83      1      1   18.7500        S  NaN
869         1       3    male  4.00      1      1   11.1333        S  NaN

### Ques 16. What are the chances of infants surviving ?

In [22]:
children = df[df["age"] <= 4]
child_chance_of_survival = surviving_children.shape[0] / children.shape[0]
child_chance_of_survival * 100

67.5

There is 67.5% chances that infant will survive.

********************************************************************************

### Ques 17. Find the teenage passengers data (age < 18) based on their gender.

In [23]:
df_group = df.groupby(by=df["sex"], as_index=False) # df_group is an iterable DataFrameGroupBy object
df_group

In [24]:
for _, df_sex in df_group:
  df_sex = df_sex[df["age"].apply(lambda x: x < 18)]
  print(df_sex)

     survived  pclass     sex    age  sibsp  parch      fare embarked deck
9           1       2  female  14.00      1      0   30.0708        C  NaN
10          1       3  female   4.00      1      1   16.7000        S    G
14          0       3  female  14.00      0      0    7.8542        S  NaN
22          1       3  female  15.00      0      0    8.0292        Q  NaN
24          0       3  female   8.00      3      1   21.0750        S  NaN
39          1       3  female  14.00      1      0   11.2417        C  NaN
43          1       2  female   3.00      1      2   41.5792        C  NaN
58          1       2  female   5.00      1      2   27.7500        S  NaN
68          1       3  female  17.00      4      2    7.9250        S  NaN
71          0       3  female  16.00      5      2   46.9000        S  NaN
84          1       2  female  17.00      0      0   10.5000        S  NaN
111         0       3  female  14.50      1      0   14.4542        C  NaN
114         0       3  fe

/var/folders/nt/3mk_43xn04n4h7tyjmwhy93w0000gp/T/ipykernel_1575/3280849714.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_sex = df_sex[df["age"].apply(lambda x: x < 18)]
/var/folders/nt/3mk_43xn04n4h7tyjmwhy93w0000gp/T/ipykernel_1575/3280849714.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_sex = df_sex[df["age"].apply(lambda x: x < 18)]


## **Great Learning :)**